netket/graph/lattice.py

```diff
@@ -104,323 +107,349 @@ def get_edges(atoms_positions, cutoff, distance_atol=cutoff_tol):
     return pairs
 
 
-def create_points(basis_vectors, extent, atom_coords, pbc):
+@dataclass
+class LatticeSite:
+    id: int
+    """Integer ID of this site"""
+    position: Array
+    """Real-space position of this site"""
+    cell_coord: Array
+    """Cell coordinates of this site"""
+    inside: bool
+    """TODO: When exactly is this needed?"""
+
+    def __repr__(self):
+        s = ", ".join(map(str, (self.id, self.cell_coord)))
+        return f"LatticeSite({s})"
+
+
+def create_sites(primitives, extent, apositions, pbc):
     shell_vec = _np.zeros(extent.size, dtype=int)
     shift_vec = _np.zeros(extent.size, dtype=int)
     # note: by modifying these, the number of shells can be tuned.
     shell_vec[pbc] = 2
     shift_vec[pbc] = 1
     ranges = tuple([list(range(ex)) for ex in extent + shell_vec])
-    atoms = []
-    cellANDlabel_to_site = {}
+    sites = []
+    cell_coord_to_site = {}
     for s_cell in itertools.product(*ranges):
         s_coord_cell = _np.asarray(s_cell) - shift_vec
         if _np.any(s_coord_cell < 0) or _np.any(s_coord_cell > (extent - 1)):
             inside = False
         else:
             inside = True
-        atom_count = len(atoms)
-        for i, atom_coord in enumerate(atom_coords):
-            s_coord_atom = s_coord_cell + atom_coord
-            r_coord_atom = _np.matmul(basis_vectors.T, s_coord_atom)
-            atoms.append(
-                {
-                    "Label": i,
-                    "cell": s_coord_cell,
-                    "r_coord": r_coord_atom,
-                    "inside": inside,
-                }
+        atom_count = len(sites)
+        for i, atom_coord in enumerate(apositions):
+            s_coord_site = s_coord_cell + atom_coord
+            r_coord_site = _np.matmul(primitives.T, s_coord_site)
+            cell_coord_site = _np.array((*s_coord_cell, i), dtype=int)
+            sites.append(
+                LatticeSite(
+                    id=None,  # to be set later, after sorting all sites
```

Comment:
```suggestion
                    id = len(sites),  # we list sites in lexicographic order
```

Reply:
The corresponding NetworkX stuff needs to be removed, but I need to leave that to you, I don't understand it

Reply:
Is this meant as another choice of "invalid id" indicator? We do change it in `__init__` a moment later anyways. (I think having an incrementing id here would be equivalent since the sort order of the sites seems to not actually be changed, despite the `__init__` code containing some sorting logic.)

Reply:
It is meant as setting `id` once and for all. The NetworkX sorting logic just gives an opportunity for things to break depending on implementation details of NetworkX, so it should be removed. The only useful purpose that could have is ensuring the labels are contiguous, but this achieves that too.

Reply:
Hm, for some reason setting the id there and touching the reordering code in `__init__` seems to break the ordering test I added in the last commit. I'm confused. :sweat_smile: